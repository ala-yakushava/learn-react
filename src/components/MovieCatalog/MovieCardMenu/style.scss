@import "../../../styles/theme";

.MovieCardMenu {
  position: absolute;
  top: 0;
  right: 0;
  display: flex;
  flex-direction: column;
  width: 200px;
  padding-top: 20px;
  padding-bottom: 10px;
  background-color: $darkest-grey;

  &_close {
    @include reset-button;
    position: absolute;
    top: 5px;
    right: 5px;
    padding: 5px;
    color: $white;
  }

  &_button {
    @include reset-button;
    display: inline-block;
    padding: 10px 20px;
    color: $white;
    text-align: left;
    background-color: $darkest-grey;
  }
}
