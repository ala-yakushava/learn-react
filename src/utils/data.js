import uniqueId from 'lodash/uniqueId';

export const sortItems = [
  {
    id: 'sort-rating',
    value: 'vote_average',
    label: 'Rating',
  },
  {
    id: 'sort-release_date',
    value: 'release_date',
    label: 'Release date',
  },
];

export const filterItems = [
  {
    id: 'filter-all',
    value: 'All',
    label: 'All',
  },
  {
    id: 'filter-documentary',
    value: 'Documentary',
    label: 'Documentary',
  },
  {
    id: 'filter-comedy',
    value: 'Comedy',
    label: 'Comedy',
  },
  {
    id: 'filter-horror',
    value: 'Horror',
    label: 'Horror',
  },
  {
    id: 'filter-crime',
    value: 'Crime',
    label: 'Crime',
  },
];

export const formField = [
  {
    id: uniqueId(),
    name: 'title',
    type: 'text',
    label: 'Title',
    placeholder: 'Title here',
  },
  {
    id: uniqueId(),
    name: 'release_date',
    type: 'date',
    label: 'Release date',
    placeholder: 'Select date',
  },
  {
    id: uniqueId(),
    name: 'poster_path',
    type: 'url',
    label: 'Movie URL',
    placeholder: 'Movie URL here',
  },
  {
    id: uniqueId(),
    name: 'genres',
    type: 'text',
    label: 'Genre',
    placeholder: 'Select genre',
  },
  {
    id: uniqueId(),
    name: 'overview',
    type: 'text',
    label: 'Overview',
    placeholder: 'Overview text goes here',
  },
  {
    id: uniqueId(),
    name: 'runtime',
    type: 'text',
    label: 'Runtime',
    placeholder: 'Runtime text goes here',
  },
];
