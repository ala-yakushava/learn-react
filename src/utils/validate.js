import * as yup from 'yup';

const MovieFormSchema = yup.object().shape({
  title: yup.string().transform((value) => value.trim()).min(2).required('Required'),
  release_date: yup.date().max(new Date(), 'Future date'),
  poster_path: yup.string().url('Invalid URL').required('Required'),
  genres: yup.string().transform((value) => value.trim()).min(3).required('Required'),
  overview: yup.string().transform((value) => value.trim()).min(10).required('Required'),
  runtime: yup.number('Invalid number').required('Required').min(0, 'Too Short!').integer(),
});

export default MovieFormSchema;
