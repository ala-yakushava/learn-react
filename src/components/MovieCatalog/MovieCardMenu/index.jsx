import React from 'react';
import PropTypes from 'prop-types';

import './style.scss';
import ModalPortal from '../../ModalPortal';
import Modal from '../../Modal';
import EditMovieForm from '../../Modal/EditMovieForm';
import DeleteMovieDialog from '../../Modal/DeleteMovieDialog';

class MovieCardMenu extends React.Component {
  constructor(props) {
    super(props);
    this.state = { 
      visible: false,
      modalType: null,
    };
  }

  handleClickOpenModal = type => () => {
    this.setState(state => ({
      visible: true,
      modalType: type,
    }));
  }

  handleClickCloseModal = () => {
    this.setState(state => ({
      visible: false,
      modalType: null,
    }));
  }

  modalBody = {
    editMovie: <EditMovieForm movieId={this.props.movieId} />,
    deleteMovie: <DeleteMovieDialog movieId={this.props.movieId} />,
  };

  render() {
    return (
      <div className="MovieCardMenu">
        <button type="button" className="MovieCardMenu_close" onClick={this.props.onClickCloseMenu}>X</button>
        <button className="MovieCardMenu_button" onClick={this.handleClickOpenModal('deleteMovie')}>Delete</button>
        <button className="MovieCardMenu_button" onClick={this.handleClickOpenModal('editMovie')}>Edit</button>
        {this.state.visible &&
          <ModalPortal>
            <Modal onClick={this.handleClickCloseModal}>
              {this.modalBody[this.state.modalType]}
            </Modal>
          </ModalPortal>
        }
      </div>
    )
  }
};

MovieCardMenu.propTypes = {
  onClickCloseMenu: PropTypes.func,
};

MovieCardMenu.defaultProps = {
 onClickCloseMenu: () => {},
};

export default MovieCardMenu;
