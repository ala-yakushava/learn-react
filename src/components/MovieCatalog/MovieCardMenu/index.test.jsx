import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import MovieCardMenu from './index';
import { id, movie } from '../../../utils/mock-data';

jest.mock('react-redux');

describe('Components - MovieCardMenu', () => {
  const handleClose = jest.fn();
  const handleRemove = jest.fn();
  const dispatch = jest.fn();

  const state = {
    movieInfo: { movie },
    loading: { loading: false },
  };

  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((element) => element);
    useSelector.mockImplementation((cb) => cb(state));
    useDispatch.mockReturnValue(dispatch);

    const modalRoot = document.createElement('div');
    modalRoot.setAttribute('id', 'modal-root');
    const body = document.querySelector('body');
    body.appendChild(modalRoot);
  });

  afterAll(() => {
    ReactDOM.createPortal.mockClear();
    useDispatch.mockClear();
  });

  test('should match snapshot', () => {
    const tree = renderer
      .create(
        <MovieCardMenu movieId={id} onClickCloseMenu={handleClose} onRemoveMovie={handleRemove} />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should open modal with EditMovieForm', () => {
    const { getByTestId, getByRole } = render(
      <MovieCardMenu movieId={id} onClickCloseMenu={handleClose} onRemoveMovie={handleRemove} />,
    );

    const leftClick = { button: 0 };
    const button = getByRole('button', { name: /Edit/i });
    userEvent.click(button, leftClick);

    const modal = getByTestId('modal');
    const dialog = getByTestId('edit-form');
    expect(modal).toContainElement(dialog);

    const closeModal = getByRole('button', { name: /x/i });
    userEvent.click(closeModal, leftClick);
    expect(modal).not.toBeInTheDocument();
  });

  test('should open modal with DeleteMovieDialog', () => {
    const { getByTestId, getByRole } = render(
      <MovieCardMenu movieId={id} onClickCloseMenu={handleClose} onRemoveMovie={handleRemove} />,
    );

    const leftClick = { button: 0 };
    const button = getByRole('button', { name: /Delete/i });
    userEvent.click(button, leftClick);

    const modal = getByTestId('modal');
    const dialog = getByTestId('delete-dialog');
    expect(modal).toContainElement(dialog);

    const closeModal = getByRole('button', { name: /x/i });
    userEvent.click(closeModal, leftClick);
    expect(modal).not.toBeInTheDocument();
  });
});
