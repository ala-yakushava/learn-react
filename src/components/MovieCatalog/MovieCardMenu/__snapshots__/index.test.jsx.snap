// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components - MovieCardMenu should match snapshot 1`] = `
<div
  className="MovieCardMenu"
  data-testid="card-menu"
>
  <button
    className="MovieCardMenu_close"
    onClick={[MockFunction]}
    type="button"
  >
    Ð¥
  </button>
  <button
    className="MovieCardMenu_button"
    onClick={[Function]}
    type="button"
  >
    Delete
  </button>
  <button
    className="MovieCardMenu_button"
    onClick={[Function]}
    type="button"
  >
    Edit
  </button>
</div>
`;
